### PR TITLE
CP-40844: Adds download source action that gets the latest source code in the production stage

### PR DESCRIPTION
**Pull request by CitrixChris**
… in the production stage.

### CODE REVIEW

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (danilo-delbusso) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:157]: - I would use the `BrandingManager`'s version of `XenCenter` to name the file.
- Also the filter should be localised since it's visible to customers. Possibly even the `-source` bit in the file name. If you don't want to localise it, we can remove "zip files" but ideally it would. Maybe check what we do in other similar cases.
- For the `InitialDirectory`, using "Downloads" as a name for the folder doesn't necessarily work for all locales. The server status report uses `Win32.GetKnownFolderPath(Win32.KnownFolders.Downloads);` to fetch the path. While I'm not the biggest fan of calling the win32 APIs directly, we already have the methods so might as well use them.
- I'd also use the object initialiser, I think it looks neater. But that's just preference so fell free to ignore:
```csharp
 var saveSourceDialog = new SaveFileDialog
            {
                FileName = "XenCenter-source",
                DefaultExt = "zip",
                Filter = "Zip files (*.zip)|*.zip|All files (*.*)|*.*",
                InitialDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),"Downloads")
            };
```

File: XenModel/Actions/Updates/DownloadSourceAction.cs
Comment (danilo-delbusso) [inline XenModel/Actions/Updates/DownloadSourceAction.cs:64]: Leftover commented out code here

**Review by danilo-delbusso — CHANGES_REQUESTED** [review]
While making the changes for the source info to be fetched as part of the client updates XML, there's some low hanging fruits that can be changed even before that 👇

File: XenModel/Actions/Updates/DownloadSourceAction.cs
Comment (danilo-delbusso) [reply XenModel/Actions/Updates/DownloadSourceAction.cs:1]: Mh I can see this being a big pain when the merge feature branch is merged though. not sure what the best way forward is, though

https://github.com/xenserver/xenadmin/pull/3152/files/194e61b988469087b94622112b399d22bf412dc0..23cea7e250aa9c64837832e82baae49f9a643eb9

File: XenModel/BrandManager.cs
Comment (kc284) [inline XenModel/BrandManager.cs:34]: The changes in BrandManager.cs and AssemblyInfo.cs are not needed since we're not introducing any new branded values. If I remember well we said we would be adding two new entries `source_url` and `source_checksum` in the XCupdates.xml?

File: XenAdmin/MainWindow.cs
Comment (kc284) [inline XenAdmin/MainWindow.cs:3280]: Aplogies if I'm missing something, but why have we added two items?

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (kc284) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:153]: It's a goog practice to construct and launch modal dialogs within a using directive so it gets disposed after use and releases OS resources.

File: XenModel/Actions/Updates/DownloadSourceAction.cs
Comment (kc284) [reply XenModel/Actions/Updates/DownloadSourceAction.cs:1]: 👍 
Or you could make the DownloadFileAction an abstract base class containing the shared bits and derive the other two from it.

**Review by kc284 — CHANGES_REQUESTED** [review]
As above.

File: XenModel/BrandManager.cs
Comment (CitrixChris) [reply XenModel/BrandManager.cs:34]: source_url yes but i tought we werent bothered with the checksum on the source as it is just the source. We can though

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (danilo-delbusso) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:170]: probably meant `FirstOrDefault` here?

> The [First<TSource>(IEnumerable<TSource>)](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.first?view=netframework-4.8.1#system-linq-enumerable-first-1(system-collections-generic-ienumerable((-0)))) method throws an exception if source contains no elements. To instead return a default value when the source sequence is empty, use the [FirstOrDefault](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.firstordefault?view=netframework-4.8.1) method.

```suggestion
            var clientVersion = Updates.ClientVersions.FirstOrDefault();
```

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (danilo-delbusso) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:35]: `using` here is not required 👀 

```suggestion
```

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (danilo-delbusso) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:154]: super nit picky but we don't use `string.Empty`:


```suggestion
            string outputPathAndFileName;
```

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (danilo-delbusso) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:197]: again nit picky but we can get rid of `return`;

```suggestion
```

**Review by danilo-delbusso — CHANGES_REQUESTED** [review]
Gonna leave it at this for now, I'd rather wait for the merge conflicts to be addressed first as they're not super straight forward and I might just be commenting on stuff that's already been fixed on `master`

File: XenModel/Messages.resx
Comment (kc284) [inline XenModel/Messages.resx:4998]: There are a lot of duplicate keys here, this can't compile.

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (kc284) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:169]: Trivial: return on a new line is clearer.

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (kc284) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:179]: I don't think the rest of this method is needed; we only save the zip, we don't need to close the application or disconnect, hence no need to check for unfinished tasks.

File: XenAdmin/MainWindow.resx
Comment (kc284) [inline XenAdmin/MainWindow.resx:3190]: No fullstop needed at the end.

File: XenAdmin/MainWindow.cs
Comment (kc284) [inline XenAdmin/MainWindow.cs:2719]: I think these two are the other way round, it's downloadSourceToolStripMenuItem that's associated with the alert, right?

Also, downloadLatestSourceToolStripMenuItem is always there even if the check does not run automatically, so I believe we need to format it (in the construcotr alongside other items) using the current version.

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (kc284) [inline XenAdmin/Alerts/Types/ClientUpdateAlert.cs:168]: This method is also called from the item under the help menu when alerts are absent, in which case it won't download the source. Maybe make the clientVersion a parameter and set it accordingly in the item click handlers.

**Review by kc284 — CHANGES_REQUESTED** [review]

File: XenAdminTests/XenAdminTests.csproj
Comment (CitrixChris) [reply XenAdminTests/XenAdminTests.csproj:166]: No idea, Visual studio must have deemed it necessary, ill revert it.

File: scripts/xenadmin-build.sh
Comment (CitrixChris) [reply scripts/xenadmin-build.sh:207]: Ah no worries good spot

File: XenAdmin/MainWindow.cs
Comment (CitrixChris) [reply XenAdmin/MainWindow.cs:2719]: Good spot on the alert-associated tooltip. 

Now that I've swapped them though I don't understand what you mean about its always there. It will be there if there is updates.clientversions has something in it. It always should be the latest version of XC so if I have the latest version I should show the button in the help menu so I can download that source. It will be hidden if the xcupdates.xml fails to download or doesn't exist.

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (CitrixChris) [reply XenAdmin/Alerts/Types/ClientUpdateAlert.cs:168]: It will as the clientVersions is the pre-processed list of updates, it only becomes an update alert if the installed version < update version

File: XenModel/BrandManager.cs
Comment (danilo-delbusso) [inline XenModel/BrandManager.cs:56]: We're not using this value in the `BrandManager`. Should we remove it?

File: XenAdmin/MainWindow.resx
Comment (danilo-delbusso) [inline XenAdmin/MainWindow.resx:1]: there's an awful lot of automated changes here that are hard to decipher like this. Is this VS22 doing its thing?

**Review by danilo-delbusso — CHANGES_REQUESTED** [review]

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (CitrixChris) [reply XenAdmin/Alerts/Types/ClientUpdateAlert.cs:172]: Interesting as that is true current but it shouldn't ever be a problem once this is merged but i should handle it better just incase

File: XenAdmin/MainWindow.cs
Comment (CitrixChris) [reply XenAdmin/MainWindow.cs:2717]: We currently pull the url from the XCupdates.xml file so we would need to fetch that before we can download anything. 
So this is to cover the event that update checks are not automatic, so the menu item would need to trigger a fetch of the updates file and pull the source url. It would default to "Download latest source" instead of having the version number auto populated, I think this is doable.

You mentioning this made me notice a bug that if the auto check for updates is disabled the menu item under help is visible without a version number so I'll fix that to be by default invisible.

File: XenModel/BrandManager.cs
Comment (CitrixChris) [reply XenModel/BrandManager.cs:56]: probably

File: XenAdmin/Alerts/Types/ClientUpdateAlert.cs
Comment (CitrixChris) [reply XenAdmin/Alerts/Types/ClientUpdateAlert.cs:172]: In fact no as the same would be true for clientversion.Name clientVersion.Version etc. Future XMLs will have a sourceURL value with this PR.

File: XenModel/Actions/Updates/DownloadFileAction.cs
Comment (CitrixChris) [reply XenModel/Actions/Updates/DownloadFileAction.cs:56]: This was done to reduce exposure of the class but still allow access.

**Comment by CitrixChris** [comment]
Needs rebasing onto master after https://github.com/xenserver/xenadmin/pull/3223  is merged

**Review by kc284 — CHANGES_REQUESTED** [review]
I have made one or two more corrections and some refactorging. This is the commit: https://github.com/kc284/xenadmin/commit/094bd183ef373de0fbe2a9f198a30a63678881a7

If you add me as collaborator to your fork, I can send a PR directly and also add comments to explain the change.

**Review by kc284 — APPROVED** [review]
I suggest we squash this when we merge it.

**Review by danilo-delbusso — APPROVED** [review]

**Pull request merged**
